### PR TITLE
Simplest pyproject.toml containing build-system only

### PR DESCRIPTION
**Pull request by ianthomas23**
This is the start of a transition to a `pyproject.toml` build system (issue #23815) and is a alternative to (or precursor to) PR #23829.

This is the only part of the `pyproject.toml` transition that is required for the `pybind11` progress to continue. So after this is merged work can continue on both `pyproject.toml` and `pybind11` independently.

### CODE REVIEW

**Comment by ianthomas23** [comment]
For C/C++ coverage to work, need a CI build using `--no-build-isolation` as otherwise the `.gcno` files are created in the isolated build environment which is then deleted.

I think we will probably need to use `--no-build-isolation` for the minimal versions CI run, so that is the obvious CI run to play with.

**Comment by ianthomas23** [comment]
This is now working as expected, and is the smallest set of changes needed to use `pyproject.toml`'s `[build-system]`. All of the GitHub Action `test` runs except for the minimum versions one now use build isolation, i.e. build in a temporary isolated virtual environment to create a wheel and that wheel is then installed in the current working environment. The minimum versions test run does not use build isolation so that it builds against the already installed minimum versions of dependencies.

File: setup.py [hunk 1]
```diff
@@ -208,8 +217,9 @@ def update_matplotlibrc(path):
 class BuildPy(setuptools.command.build_py.build_py):
     def run(self):
         super().run()
-        update_matplotlibrc(
-            Path(self.build_lib, "matplotlib/mpl-data/matplotlibrc"))
+        if not self.editable_mode:
```
Comment (ianthomas23) [inline setup.py:220]: It surprised me to discover that when you build the `main` branch in editable mode this `run` function is not called. This branch does call the `run` function so here I am explicitly avoiding the `update_matplotlibrc` call when in editable mode to give the same behaviour as `main` branch.

**Review by oscargus — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (timhoffm) [reply setup.py:220]: Editable (via pip/setuptools) is quite different from a regular build. I'm not an expert there, but have the feeling that if in doubt expectations on editable installs do not hold. Possibly [pep660](https://peps.python.org/pep-0660/) will change that.

**Comment by tacaswell** [comment]
Lets get this in ASAP so we can shake out any issues with it.

**Pull request merged**